base/log_level.go

```diff
@@ -0,0 +1,67 @@
+package base
+
+import (
+	"errors"
+	"sync/atomic"
+)
+
+// LogLevel is used to represent a log level.
+type LogLevel uint32
+
+const (
+	// LEVEL_NONE disables all logging
+	LEVEL_NONE LogLevel = iota
+	// LEVEL_ERROR enables only error logging.
+	LEVEL_ERROR
+	// LEVEL_WARN enables warn, and error logging.
+	LEVEL_WARN
+	// LEVEL_INFO enables info, warn, and error logging.
+	LEVEL_INFO
+	// LEVEL_DEBUG enables all logging.
+	LEVEL_DEBUG
+)
+
+var (
+	logLevelNames      = []string{"none", "error", "warn", "info", "debug"}
+	logLevelNamesPrint = []string{"NON", "ERR", "WRN", "INF", "DBG"}
+
+	ErrInvalidLogLevel = errors.New("invalid log level")
+)
+
+func (l *LogLevel) Set(newLevel LogLevel) {
+	atomic.StoreUint32((*uint32)(l), uint32(newLevel))
+}
+
+// Enabled returns true if the log key is enabled.
+func (l *LogLevel) Enabled(logLevel LogLevel) bool {
+	return atomic.LoadUint32((*uint32)(l)) >= uint32(logLevel)
+}
+
+// LogLevelName returns the string representation of a log level (e.g. "debug" or "warn")
+func LogLevelName(logLevel LogLevel) string {
+	// No lock required to read concurrently, as long as nobody writes to logLevelNames.
+	return logLevelNames[logLevel]
+}
+
+// logLevelNamePrint returns the string value to print for a log level (e.g. "DBG" or "WRN")
+func logLevelNamePrint(logLevel LogLevel) string {
+	// No lock required to read concurrently, as long as nobody writes to logLevelNames.
+	return logLevelNamesPrint[logLevel]
+}
+
+func (l *LogLevel) MarshalText() (text []byte, err error) {
+	if l == nil {
+		return nil, ErrInvalidLogLevel
+	}
+	return []byte(LogLevelName(*l)), nil
+}
+
+func (l *LogLevel) UnmarshalText(text []byte) error {
+	for i, name := range logLevelNames {
+		if name == string(text) {
+			*l = LogLevel(i)
```

Comment:
Same question as earlier - should we be using atomic to write LogLevel during unmarshalling?